src/main/java/org/testng/internal/TestResult.java

```diff
@@ -320,4 +336,100 @@ public boolean wasRetried() {
   public void setWasRetried(boolean wasRetried) {
     this.m_wasRetried = wasRetried;
   }
+
+  public List<ITestNGMethod> getSkipCausedBy() {
+    if (this.m_status != SKIP) {
+      return Collections.unmodifiableList(skippedDueTo);
+    }
+    if (skipAnalysed) {
+      return Collections.unmodifiableList(skippedDueTo);
+    }
+    skipAnalysed = true;
+    //check if there were any config failures
+    Set<ITestResult> skippedConfigs = m_context.getFailedConfigurations().getAllResults();
+    for (ITestResult skippedConfig : skippedConfigs) {
+      if (isGlobalFailure(skippedConfig) || isRelated(skippedConfig) ) {
+        //If there's a failure in @BeforeTest/@BeforeSuite/@BeforeClass
+        //then the reason is most often just one method.
+        skippedDueTo.add(skippedConfig.getMethod());
+        return Collections.unmodifiableList(skippedDueTo);

```

Comment:
Is there anything wrong with doing this? I would like to understand more. Exit when a condition happens is what I prefer more, rather than reducing the number of exits and having a generic code. Maybe a programming style even. I would be happy to change, if you could please let me know the pitfalls with that thought process.

Reply:
My problem is that we are doing a functional assumption and we may miss something in case of we have wrong.
The behavior may change and the assumption too?
It will be more difficult to find the issue later.

Instead, iterating on all is not wrong (maybe it is?), reduce the code and the complexity.

Reply:
@juherr - Ok. That makes sense. Fixed